File: backend/app/schemas/broker.py
from pydantic import BaseModel, EmailStr, ConfigDict
from typing import Optional, Dict, Any
from datetime import datetime
from uuid import UUID
import json


class BrokerBase(BaseModel):
    user_id: UUID
    nickname: Optional[str] = None
    type: str


class BrokerConnect(BrokerBase):
    name: str
    password: str


class BrokerAdd(BrokerBase):
    user_broker_id: str
    access_token: str
    expire_in: int


class BrokerInfo(BrokerBase):
    id: UUID
    last_sync: datetime
    status: bool
    user_broker_id: str
    live: Optional[int] = None
    paper: Optional[int] = None
    enable: Optional[int] = None


class SummarySubBrokers(BaseModel):
    live: Optional[int] = None
    paper: Optional[int] = None
    enable: Optional[int] = None


class RespondTradoVate(BaseModel):
    accessToken: str
    mdAccessToken: str
    userId: str


class BrokerFilter(BaseModel):
    id: Optional[UUID] = None
    user_id: Optional[UUID] = None
    nickname: Optional[str] = None
    type: Optional[str] = None
    status: Optional[bool] = None
    user_broker_id: Optional[str] = None


class SubBrokerAdd(BaseModel):
    user_id: UUID
    broker_account_id: UUID
    user_broker_id: str
    sub_account_id: str
    nickname: Optional[str] = None
    sub_account_name: str
    type: str
    account_type: str
    is_demo: bool
    status: bool


class SubBrokerInfo(SubBrokerAdd):
    id: UUID
    last_sync: datetime
    is_active: bool
    broker_account_id: Optional[int] = None


class SubBrokerInfoPlus(SubBrokerInfo):
    balance: float    


class SubBrokerFilter(BaseModel):
    id: Optional[UUID] = None
    user_id: Optional[UUID] = None
    user_broker_id: Optional[str] = None
    sub_account_id: Optional[str] = None
    nickname: Optional[str] = None
    sub_account_name: Optional[str] = None
    type: Optional[str] = None
    is_demo: Optional[bool] = None
    status: Optional[bool] = None
    is_active: Optional[bool] = None


class BrokerChange(BaseModel):
    id: UUID
    nickname: Optional[str] = None
    status: Optional[bool] = None


class SubBrokerChange(BaseModel):
    id: UUID
    nickname: Optional[str] = None
    is_active: Optional[bool] = None

class SubBrokerSumary(BaseModel):
    id: UUID
    nickname: str
    sub_account_name: str
    qty: int

class SubBrokerSumaryForGet(BaseModel):
    id: UUID
    nickname: str
    sub_account_name: str

class SubBrokersWithQty(BaseModel):
    id: UUID
    qty: int
    
class ExitPosition(BaseModel):
    accountId: int
    action: str
    symbol: str
    orderQty: int
    orderType: str
    isAutomated: bool

File: frontend/src/types/broker.ts
export interface BrokerFilter {
  id?: string;
  user_id?: string;
  user_broker_id?: string;
  nickname?: string;
  type?: string;
  status?: boolean;
}

export interface BrokerInfo {
  id: string;
  user_id: string;
  user_broker_id: string;
  nickname: string;
  type: string;
  last_sync: Date;
  status: boolean;
  live: number;
  paper: number;
  enable: number;
}

export interface SubBrokerInfo {
  id: string;
  user_id: string;
  user_broker_id: string;
  sub_account_id: string;
  nickname: string;
  sub_account_name: string;
  type: string;
  account_type: string;
  last_sync: Date;
  is_demo: boolean;
  status: boolean;
  is_active: boolean;
  balance?: number;
}

export interface SubBroker {
  id: string;
  user_id: string;
  user_broker_id: string;
  sub_account_id: string;
  nickname: string;
  sub_account_name: string;
  type: string;
  account_type: string;
  last_sync: Date;
  is_demo: boolean;
  status: boolean;
  is_active: boolean;
}

export interface SubBrokerSummary {
  id: string;
  nickname: string;
  sub_account_name: string;
  qty: number;
}

export interface SubBrokerSummaryForGet {
  id: string;
  nickname: string;
  sub_account_name: string;
}

export interface SubBrokerFilter {
  id?: string;
  user_id?: string;
  user_broker_id?: string;
  sub_account_id?: string;
  nickname?: string;
  sub_account_name?: string;
  type?: string;
  is_demo?: boolean;
  status?: boolean;
  is_active?: boolean;
}

export interface BrokerChange {
  id: string;
  nickname?: string;
  status?: boolean;
}

export interface SubBrokerChange {
  id: string;
  nickname?: string;
  is_active?: boolean;
}

export interface TradeDate {
  year: number;
  month: number;
  day: number;
}
export interface TradovatePositionListResponse {
  id: number;
  accountId: number;
  accountNickname: string;
  symbol: string;
  netPos: number;
  netPrice: number;
  bought: number;
  boughtValue: number;
  sold: number;
  soldValue: number;
  accountDisplayName: string;
}

export interface TradovateOrderListResponse {
  id: number;
  accountId: number;
  accountNickname: string;
  contractId: number;
  timestamp: Date;
  action: string;
  ordStatus: string;
  executionProviderId: number;
  archived: boolean;
  external: boolean;
  admin: boolean;
  symbol: string;
  price: number;
  accountDisplayName: string;
}

export interface TradovateProductItemResponse {
  id: number;
  name: string;
  currencyId: number;
  productType: string;
  description: string;
  exchangeId: number;
  exchangeChannelId: number;
  contractGroupId: number;
  riskDiscountContractGroupId: number;
  status: string;
  months: string;
  valuePerPoint: number;
  priceFormatType: string;
  priceFormat: number;
  tickSize: number;
  allowProviderContractInfo: boolean;
  isMicro: boolean;
  marketDataSource: string;
  lookupWeight: number;
  hasReplay: boolean;
  settlementMethod: string;
}

export interface TradovateContractMaturityItemResponse {
  id: number;
  productId: number;
  expirationMonth: number;
  expirationDate: Date;
  archived: boolean;
  seqNo: number;
  isFront: boolean;
}

export interface TradovateContractItemResponse {
  id: number;
  name: string;
  contractMaturityId: number;
  status: string;
  providerTickSize: number;
}

export interface TradovateAccountsResponse {
  id: number;
  accountId: number;
  accountNickname: string;
  timestamp: Date;
  currencyId: number;
  amount: number;
  realizedPnL: number;
  weekRealizedPnL: number;
  archived: boolean;
  amountSOD: number;
  accountDisplayName: string;
}

export interface ExitPostion {
  accountId: number;
  action: string;
  symbol: string;
  orderQty: number;
  orderType: string;
  isAutomated: boolean;
}

export interface SubBrokerForCreate {
  id: string;
  qty: number;
}